### PR TITLE
eGLD-in-multiTransfer

### PR DESCRIPTION
**Pull request by sasurobert**

### CODE REVIEW

File: builtInFunctions/multiESDTNFTTransfer_test.go [hunk 1]
```diff
@@ -1193,3 +1193,221 @@ func runMultiTransfer(t *testing.T, isScToScEventLogEnabled bool) (*vmcommon.VMO
...
+			return flag != EGLDInESDTMultiTransferFlag
+		},
+	}
+
+	sender, err := multiTransfer.accounts.LoadAccount(vmInput.CallerAddr)
+	require.Nil(t, err)
+	destination, err := multiTransfer.accounts.LoadAccount(vmInput.Arguments[0])
+	require.Nil(t, err)
+
+	_, err = multiTransfer.ProcessBuiltinFunction(sender.(vmcommon.UserAccountHandler), destination.(vmcommon.UserAccountHandler), vmInput)
```
Comment (mariusmihaic) [inline builtInFunctions/multiESDTNFTTransfer_test.go:1267]: Can we check which error more specific?

File: builtInFunctions/multiESDTNFTTransfer_test.go [hunk 2]
```diff
@@ -1193,3 +1193,221 @@ func runMultiTransfer(t *testing.T, isScToScEventLogEnabled bool) (*vmcommon.VMO
...
+	_, _ = multiTransfer.accounts.Commit()
+
+	// reload accounts
+	sender, err = multiTransfer.accounts.LoadAccount(vmInput.CallerAddr)
+	require.Nil(t, err)
+	destination, err = multiTransfer.accounts.LoadAccount(vmInput.Arguments[0])
+	require.Nil(t, err)
+
+	testNFTTokenShouldExist(t, multiTransfer.marshaller, sender, []byte("token1"), 1, big.NewInt(2)) // 3 initial - 1 transferred
+	testNFTTokenShouldExist(t, multiTransfer.marshaller, destination, []byte("token1"), 1, big.NewInt(4))
```
Comment (mariusmihaic) [inline builtInFunctions/multiESDTNFTTransfer_test.go:1300]: These values are honestly pretty tricky to track unless you know exactly what `createSetupForMultiTransferWithEGLD` does and how it initializes balances. 

Perhaps we can add initial balances to accounts as config struct in `createSetupForMultiTransferWithEGLD` or add some simple comment for that func on how it initializes balances

File: builtInFunctions/multiESDTNFTTransfer_test.go [hunk 3]
```diff
@@ -1193,3 +1193,221 @@ func runMultiTransfer(t *testing.T, isScToScEventLogEnabled bool) (*vmcommon.VMO
...
+	require.Equal(t, vmcommon.Ok, vmOutput.ReturnCode)
+
+	_ = multiTransferSenderShard.accounts.SaveAccount(sender)
+	_, _ = multiTransferSenderShard.accounts.Commit()
+
+	// reload sender account
+	sender, err = multiTransferSenderShard.accounts.LoadAccount(senderAddress)
+	require.Nil(t, err)
+
+	testNFTTokenShouldExist(t, multiTransferSenderShard.marshaller, sender, token1, tokenNonce, big.NewInt(2)) // 3 initial - 1 transferred
```
Comment (mariusmihaic) [inline builtInFunctions/multiESDTNFTTransfer_test.go:1382]: Can we also check the balance?

File: builtInFunctions/multiESDTNFTTransfer_test.go [hunk 4]
```diff
@@ -1193,3 +1193,221 @@ func runMultiTransfer(t *testing.T, isScToScEventLogEnabled bool) (*vmcommon.VMO
...
+func TestESDTNFTMultiTransfer_ProcessBuiltinFunctionOnCrossShardsWithEGLD(t *testing.T) {
+	t.Parallel()
+
+	payableHandler := &mock.PayableHandlerStub{
+		IsPayableCalled: func(address []byte) (bool, error) {
+			return true, nil
+		},
+	}
+
+	multiTransferSenderShard := createESDTNFTMultiTransferWithMockArguments(1, 2, &mock.GlobalSettingsHandlerStub{})
```
Comment (mariusmihaic) [inline builtInFunctions/multiESDTNFTTransfer_test.go:1323]: Is there any way to put all this setup code in a func similar to `createSetupForMultiTransferWithEGLD` for cross shard?

File: builtInFunctions/multiESDTNFTTransfer_test.go [hunk 5]
```diff
@@ -1193,3 +1193,221 @@ func runMultiTransfer(t *testing.T, isScToScEventLogEnabled bool) (*vmcommon.VMO
...
+	require.Equal(t, sender.(vmcommon.UserAccountHandler).GetBalance(), big.NewInt(2))
+	require.Equal(t, destination.(vmcommon.UserAccountHandler).GetBalance(), big.NewInt(1))
+
+	funcName, args := extractScResultsFromVmOutput(t, vmOutput)
+
+	scCallFunctionAsHex := hex.EncodeToString([]byte("functionToCall"))
+	scCallArg := hex.EncodeToString([]byte("arg"))
+	assert.Equal(t, scCallFunctionAsHex, funcName)
+	require.Equal(t, 1, len(args))
+	require.Equal(t, []byte(scCallArg), args[0])
```
Comment (mariusmihaic) [inline builtInFunctions/multiESDTNFTTransfer_test.go:1311]: This code is duplicated in the other test as well. Can we extract it into a more generic func?

File: builtInFunctions/changeOwnerAddress.go [hunk 6]
```diff
@@ -114,7 +114,11 @@ func (c *changeOwnerAddress) addOutputTransferToVmOutputForCallThroughSC(acntDst
 		vmOutput)
 }
 
-func computeGasRemaining(snd vmcommon.UserAccountHandler, gasProvided uint64, gasToUse uint64) uint64 {
+func computeGasRemaining(snd vmcommon.UserAccountHandler, gasProvided uint64, gasToUse uint64, noGasUse bool) uint64 {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply builtInFunctions/changeOwnerAddress.go:117]: good idea. thanks.

File: builtInFunctions/esdtTransfer.go [hunk 7]
```diff
@@ -404,3 +405,10 @@ func (e *esdtTransfer) SetPayableChecker(payableHandler vmcommon.PayableChecker)
 func (e *esdtTransfer) IsInterfaceNil() bool {
 	return e == nil
 }
+
+func noGasUseIfReturnCallAfterError(enableEpochsHandler vmcommon.EnableEpochsHandler, vmInput *vmcommon.ContractCallInput) bool {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply builtInFunctions/esdtTransfer.go:409]: would leave as it is. I wanted only to ensure backward compatibility here.

File: builtInFunctions/esdtNFTTransfer.go [hunk 8]
```diff
@@ -386,11 +387,14 @@ func (e *esdtNFTTransfer) createNFTOutputTransfers(
 			return err
 		}
 
-		gasForTransfer := uint64(len(marshaledNFTTransfer)) * e.gasConfig.DataCopyPerByte
-		if gasForTransfer > vmOutput.GasRemaining {
-			return ErrNotEnoughGas
+		if !noGasUse {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply builtInFunctions/esdtNFTTransfer.go:390]: changed to skiGasUse

File: builtInFunctions/esdtNFTTransfer.go [hunk 9]
```diff
@@ -460,12 +460,12 @@ func addNFTTransferToVMOutput(
 	}
 	outTransfer := vmcommon.OutputTransfer{
 		Index:         index,
-		Value:         big.NewInt(0),
+		Value:         big.NewInt(0).Set(vmInput.CallValue),
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply builtInFunctions/esdtNFTTransfer.go:463]: yes. it is backward compatible. The flags ensure that code does not reach this place if flag is not set.

File: builtInFunctions/multiESDTNFTTransfer.go [hunk 10]
```diff
@@ -23,9 +23,11 @@ type esdtNFTMultiTransfer struct {
 	gasConfig      vmcommon.BaseOperationCost
 	mutExecution   sync.RWMutex
 	rolesHandler   vmcommon.ESDTRoleHandler
+	baseTokenID           []byte
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply builtInFunctions/multiESDTNFTTransfer.go:26]: github.

File: builtInFunctions/multiESDTNFTTransfer.go [hunk 11]
```diff
@@ -200,6 +203,15 @@ func (e *esdtNFTMultiTransfer) ProcessBuiltinFunction(
 		} else {
 			transferredValue := big.NewInt(0).SetBytes(vmInput.Arguments[tokenStartIndex+2])
 			value.Set(transferredValue)
+
+			if bytes.Equal(e.baseTokenID, tokenID) {
+				err = acntDst.AddToBalance(transferredValue)
+				if err != nil {
+					return nil, err
+				}
+				continue
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply builtInFunctions/multiESDTNFTTransfer.go:212]: done. good suggesiton.

File: builtInFunctions/multiESDTNFTTransfer.go [hunk 12]
```diff
@@ -278,8 +290,9 @@ func (e *esdtNFTMultiTransfer) processESDTNFTMultiTransferOnSenderShard(
 		return nil, fmt.Errorf("%w, invalid number of arguments", ErrInvalidArguments)
 	}
 
+	noGasUse := noGasUseIfReturnCallAfterError(e.enableEpochsHandler, vmInput)
 	multiTransferCost := numOfTransfers * e.funcGasCost
-	if vmInput.GasProvided < multiTransferCost {
+	if vmInput.GasProvided < multiTransferCost && !noGasUse {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply builtInFunctions/multiESDTNFTTransfer.go:295]: changed to skipGasUse

File: builtInFunctions/multiESDTNFTTransfer.go [hunk 13]
```diff
@@ -503,11 +558,13 @@ func (e *esdtNFTMultiTransfer) createESDTNFTOutputTransfers(
 					return err
 				}
 
-				gasForTransfer := uint64(len(marshaledNFTTransfer)) * e.gasConfig.DataCopyPerByte
-				if gasForTransfer > vmOutput.GasRemaining {
-					return ErrNotEnoughGas
+				if !noGasUse {
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply builtInFunctions/multiESDTNFTTransfer.go:561]: skipGasUse

File: builtInFunctions/multiESDTNFTTransfer_test.go (hunk 1 above)
Comment (sasurobert) [reply builtInFunctions/multiESDTNFTTransfer_test.go:1267]: done

File: builtInFunctions/multiESDTNFTTransfer_test.go (hunk 2 above)
Comment (sasurobert) [reply builtInFunctions/multiESDTNFTTransfer_test.go:1300]: only a comment is enough here.

File: builtInFunctions/multiESDTNFTTransfer_test.go (hunk 3 above)
Comment (sasurobert) [reply builtInFunctions/multiESDTNFTTransfer_test.go:1382]: done

File: builtInFunctions/multiESDTNFTTransfer_test.go (hunk 4 above)
Comment (sasurobert) [reply builtInFunctions/multiESDTNFTTransfer_test.go:1323]: yes, it would mean to recheck multiple tests.

**Review by mariusmihaic — DISMISSED** [review]

**Review by BeniaminDrasovean — APPROVED** [review]

**Review by mariusmihaic — APPROVED** [review]

**Pull request merged**